File: zennode/task1.cpp
#include<bits/stdc++.h>

using namespace std;

class Shop
{
    public:
    int price_A,price_B,price_C,gift_wrap_fee,ship_fee_perPackage;

    Shop(){
        price_A = 20;
        price_B = 40;
        price_C = 50;

        gift_wrap_fee = 1;
        ship_fee_perPackage = 5;
    }

    double flat_10_discount(double totalPrice){
        if(totalPrice > 200){
            return totalPrice - 10;
        }
        return totalPrice;
    }

    double bulk_5_discount(int A, int B, int C, double totalPrice){
        if(A>10){
            totalPrice -= (A*price_A)*0.05;
        }
        if(B>10){
            totalPrice -= (B*price_B)*0.05;
        }
        if(C>10){
            totalPrice -= (C*price_C)*0.05;
        }
        return totalPrice;

    }

    double bulk_10_discount(double totalPrice, double totalUnit){
        if(totalUnit > 20){
            totalPrice -= (totalPrice*0.1);
        }
        return totalPrice;
    }

    double tiered_50_discount(int A, int B, int C, double totalPrice,double totalUnit){
        if(totalUnit > 30){
            if(A > 15){
            totalPrice -= (A-15)*price_A*0.50;
        }
        if(B > 15){
            totalPrice -= (B-15)*price_B*0.50;
        }
        if(C > 15){
            totalPrice -= (C-15)*price_C*0.50;
        }

        }
        return totalPrice;
    }

    double getShipCharge(double totalUnit){
        return ceil(totalUnit/10)*ship_fee_perPackage;
    }

};

int main(){
    Shop S;
    int quant_A,quant_B,quant_C;
    int wrapReq_A,wrapReq_B,wrapReq_C;

    cout<<"Enter quantity for Product A: ";
    cin >> quant_A;

    cout<<"Enter quantity for Product B: ";
    cin >> quant_B;

    cout<<"Enter quantity for Product C: ";
    cin >> quant_C;

    cout << "Wrap Product A as a gift? (If yes , press 1 else 0) : ";
    cin >>  wrapReq_A;

    cout << "Wrap Product B as a gift? (If yes , press 1 else 0): ";
    cin >>  wrapReq_B;

    cout << "Wrap Product C as a gift? (If yes , press 1 else 0): ";
    cin >>  wrapReq_C;

    int pA_total = quant_A* S.price_A;
    int pB_total = quant_B* S.price_B;
    int pC_total = quant_C* S.price_C;

    double totalUnit = quant_A + quant_B + quant_C;
    double totalPrice = pA_total + pB_total + pC_total;
    double shipping_fee = S.getShipCharge(totalUnit);

    int wrap_fee = 0;
    if(wrapReq_A){
        wrap_fee += quant_A;
    }
    if(wrapReq_B){
        wrap_fee += quant_B;
    }
    if(wrapReq_C){
        wrap_fee += quant_C;
    }

    set<pair<double,string>>disc;

    disc.insert({S.flat_10_discount(totalPrice),"flat_10_discount"});
    disc.insert({S.bulk_5_discount(quant_A,quant_B,quant_C,totalPrice),"bulk_5_discount"});
    disc.insert({S.bulk_10_discount(totalPrice,totalUnit),"bulk_10_discount"});
    disc.insert({S.tiered_50_discount(quant_A,quant_B,quant_C,totalPrice,totalUnit),"tiered_50_discount"});

    pair<double,string> bestDisc = (*(disc.begin())).first == totalPrice ? make_pair(totalPrice,"No discount") : *(disc.begin());

     cout << "Product A Quantity: " << quant_A << " Total amount: $" << pA_total << endl;
    cout << "Product B Quantity: " << quant_B << " Total amount: $" << pB_total << endl;
    cout << "Product C Quantity: " << quant_C << " Total amount: $" << pC_total << endl;
    cout << "Subtotal: $" << totalPrice << endl;
    cout << "Discount Scheme: " << bestDisc.second << endl;
    cout << "Discout Amount: $" << totalPrice - bestDisc.first << endl;
    cout << "The Shipping Fee: $" << shipping_fee << endl;
    cout << "Gift Wrap Fee: $" << wrap_fee << endl;
    cout << "Total $" << bestDisc.first + wrap_fee + shipping_fee << endl;



}

File: zennode/task1.js
let price = [20.0, 40.0, 50.0];
let A, B, C;
let a1, b1, c1;
let ten = 10.0, zero = 0.0;

console.log("Enter the Quantity of product A");
A = parseFloat(prompt());
console.log("Do you want to wrap product A as a Gift (Y/N): ");
a1 = prompt();

console.log("Enter the Quantity of product B");
B = parseFloat(prompt());
console.log("Do you want to wrap product B as a Gift (Y/N): ");
b1 = prompt();

console.log("Enter the Quantity of product C");
C = parseFloat(prompt());
console.log("Do you want to wrap product C as a Gift (Y/N): ");
c1 = prompt();

let cartTotal = A * price[0] + B * price[1] + C * price[2];
let giftTotal = (a1 === 'Y' ? A : 0) + (b1 === 'Y' ? B : 0) + (c1 === 'Y' ? C : 0);
let shippingCost = Math.ceil((A + B + C ) / 10) * 5;

let totalItems = A + B + C;

let flat_10_discount = 0, bulk_5_discount = 0, bulk_10_discount = 0, tiered_50_discount = 0;

if (cartTotal > 200) {
    flat_10_discount = 10;
}

bulk_5_discount = (A > 10 ? price[0] * A * 0.05 : 0) + (B > 10 ? price[1] * B * 0.05 : 0) + (C > 10 ? price[2] * C * 0.05 : 0);

bulk_10_discount = ((A + B + C) > 20 ? cartTotal * 0.1 : 0);

if ((A + B + C) > 30) {
    tiered_50_discount += (A > 15 ? (A - 15) * (price[0] / 2) : 0);
    tiered_50_discount += (B > 15 ? (B - 15) * (price[1] / 2) : 0);
    tiered_50_discount += (C > 15 ? (C - 15) * (price[2] / 2) : 0);
}

let mxDiscount = 0;
let discountType = "None";

if (mxDiscount < flat_10_discount) {
    mxDiscount = flat_10_discount;
    discountType = "flat_10_discount";
}

if (mxDiscount < bulk_5_discount) {
    mxDiscount = bulk_5_discount;
    discountType = "bulk_5_discount";
}

if (mxDiscount < bulk_10_discount) {
    mxDiscount = bulk_10_discount;
    discountType = "bulk_10_discount";
}

if (mxDiscount < tiered_50_discount) {
    mxDiscount = tiered_50_discount;
    discountType = "tiered_50_discount";
}

let finalCost = cartTotal - mxDiscount;
finalCost += giftTotal;
finalCost += shippingCost;

console.log(`Product Name: A  Quantity : ${A}  Cost : ${A * price[0]}`);
console.log(`Product Name: B  Quantity : ${B}  Cost : ${B * price[1]}`);
console.log(`Product Name: C  Quantity : ${C}  Cost : ${C * price[2]}`);
console.log(`SubTotal : ${cartTotal}`);
console.log(`The discount name applied & the discount amount : ${discountType}  ${mxDiscount}`);
console.log(`The shipping fee & the gift wrap fee : ${giftTotal + shippingCost}`);
console.log(`Total Cost: ${finalCost}`);
